Boundary extraction

sriva-e opened the pull request:
changed inference.py to boundary_extraction.py
removed "binarize_this" and made one function for binarzing i.e "convert_binary"
added Otsu and Triangle option for threshold value inside "covert_binary" function

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,39 @@
+    output_dir = "C:/Users/Sunil/PycharmProjects/pythonProject/edema-quantification/dataset/output/boundary"
+    model_name = "Unet++"
+    threshold = 'Triangle'
+
```
@sriva-e 

Add "threshold_value"

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,39 @@
+    dataset_dir = "C:/Users/Sunil/PycharmProjects/pythonProject/edema-quantification/dataset/data"
+    output_dir = "C:/Users/Sunil/PycharmProjects/pythonProject/edema-quantification/dataset/output/boundary"
+    model_name = "Unet++"
+    threshold = 'Triangle'
```
Rename this param to "method"

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,39 @@
+
+
+if __name__ == '__main__':
+    dataset_dir = "C:/Users/Sunil/PycharmProjects/pythonProject/edema-quantification/dataset/data"
```
Add argparse parameters

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,84 @@
+    )
+
+    for img_path in tqdm(img_paths, desc='Boundary extraction', unit=' images'):
+        # img_path = img_dir + '/' + img_path
```
@sriva-e

Never use this way of concatenation because developers frequently use different operating systems. That's why img_path will not be correct in some systems, for instance on Windows in this particular case. Use **os** and **pathlib** concatenation methods instead.

Remove this line of code.

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,84 @@
+    for img_path in tqdm(img_paths, desc='Boundary extraction', unit=' images'):
+        # img_path = img_dir + '/' + img_path
+        mask = model(img_path)
+        cv2.imwrite(f'dataset/output/masks/mask_{img_path}.png', mask)
```
@sriva-e 

Save masks based on the **output_dir** variable rather than a manual path 'dataset/output/masks.......'

Also, output_dir is not being created at this line of code, so the error is thrown. This is why create **output_dir** before the for loop.

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,84 @@
+        # img_path = img_dir + '/' + img_path
+        mask = model(img_path)
+        cv2.imwrite(f'dataset/output/masks/mask_{img_path}.png', mask)
+        morph = MorphologicalTransformations(
```
@sriva-e 

Revise **MorphologicalTransformations** because you should not create its object every iteration of the _for loop_. You have to create the **morph** before the loop and then use its methods in the loop. In that regard, you have to pass **thresh_method** and **thresh_val** to the class constructor, while the binary method should process **mask**.

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,84 @@
+        binarized_mask = morph.binary(thresh_method, thresh_val)
+        boundary = morph.extract_boundary(binarized_mask)
+        image_bound = morph.visualize_boundary(img_path, boundary)
+        filename = os.path.split(img_path)[-1]
```
@sriva-e 

Usage of **pathlib** is preferable because that library is much stronger. Please try to use it in order to get **img_name**

ViacheslavDanilov commented on boundary_extraction.py:
```diff
@@ -0,0 +1,84 @@
+        boundary = morph.extract_boundary(binarized_mask)
+        image_bound = morph.visualize_boundary(img_path, boundary)
+        filename = os.path.split(img_path)[-1]
+        cv2.imwrite(os.path.join(output_dir, filename), image_bound)
```
@sriva-e 

Create an additional variable **mask_path** before this line where you may initialize the output path

Keep in mind, one line of code - one operation

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+class MorphologicalTransformations:
+    def __init__(
+            self,
+            image_file: str,
```
@sriva-e 

Revise **MorphologicalTransformations** because you should not create its object every iteration of the _for loop_. You have to create **morph** before the loop and then use its methods in the loop. In that regard, you have to pass **thresh_method** and **thresh_val** to the class constructor, while the **binary** method should process mask.

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+    ):
+
+        self.image_file = image_file
+        self.MAX_PIXEL = 255
```
@sriva-e 

I still don't understand why we need **self.MAX_PIXEL** and **self.MIN_PIXEL.** I think we may remove them.

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+        self.MAX_PIXEL = 255
+        self.MIN_PIXEL = 0
+
+    def read_image(self):
```
@sriva-e 

Why do we have to read images, if we assume that the input of the binary method is ndarray image?

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+        image_src = cv2.imread(self.image_file, 0)
+        return image_src
+
+    def binary(
+            self,
+            thresh_method: str = 'otsu',
+            thresh_val: float = None,
+    ):
```
@sriva-e 

1. Rename this method to binarize
2. Input of the method should be an image rather than **thresh_method** and **thresh_val**

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+        assert thresh_method in ['otsu', 'triangle', 'manual'], f'Invalid thresh_method: {thresh_method}'
+
+        if thresh_method == 'otsu':
+            threshold_value, _ = cv2.threshold(image_src, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
```
@sriva-e 

Does **_** (underscore variable) output binary image? If so, can we use it?

sriva-e replied:
![image](https://user-images.githubusercontent.com/89933560/184637532-fc7ade53-52c8-45e1-b5a8-9f23e5957e43.png)

This is what " _" outputs , it's a binary image

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+        else:
+            logging.warning(f'Invalid threshold')
+
+        initial_conv = np.where((image_src <= threshold_value), image_src, color_1)
+        final_conv = np.where((initial_conv > threshold_value), initial_conv, color_2)
```
@sriva-e 

Is it an optimal solution?

ViacheslavDanilov commented on tools/utils.py:
```diff
@@ -263,3 +266,95 @@ def normalize_image(
+        boundary = np.expand_dims(boundary, axis=-1)
+        res_bound = cv2.resize(boundary, dsize=(1024, 1024), interpolation=cv2.INTER_CUBIC)
+
+        image = Image.open(image_src)
+        image = image.resize((1024, 1024), Image.ANTIALIAS)
+        lung = np.expand_dims(image, axis=-1)
+        dst = cv2.addWeighted(res_bound, alpha, lung, beta, 0.0, dtype=cv2.CV_64F)
+        backtorgb = cv2.cvtColor(dst.astype(np.uint8), cv2.COLOR_GRAY2RGB)
```
@sriva-e 

Why are we using two image processing libraries here: opencv and PIL? I would recommend using one of them, for instance OpenCV. It can do both reading images and resizing them afterward.

ViacheslavDanilov reviewed (changes requested):
Changes needed

sriva-e commented:
--fix 
- Renamed binary to **binarize**, removed thresh_method and thresh_val as its parameters, input of this method is now an image path.

- Used os module to get img_path

- Saved masks based on output_dir
- 
- Created an extra variable mask_path which initializes output path for masks

- Revised **MorphologicalTransformations** such that object is created outside the loop, method(binarize) is called in the loop, passed thresh_method and thresh_val to class constructor and mask_path to binarize function 

- Using path lib to find img_name 

- Removed class BoundaryDataset and Should I remove function normalize_image also? I tried running my code without it and 
it seems to be working , Please let me know why is it needed. 

 - Removed self.MAX_PIXEL and self.MIN_PIXEL
 
- Input of binarize method is now mask_path , so we are using method read_image to get a ndarray

- “_” associated with code that calculates threshold_value gives a binary image 

- Using OpenCV instead of PIL under Visualize boundary function to read image and resizing it

ViacheslavDanilov reviewed (approved):
LGTM
